pkg/dao/repository_configs.go

```diff
@@ -85,13 +85,24 @@ func (r repositoryConfigDaoImpl) Create(ctx context.Context, newRepoReq api.Repo
 		return api.RepositoryResponse{}, errors.New("Creating of Red Hat repositories is not permitted")
 	}
 
+	if newRepoReq.Origin == nil || *newRepoReq.Origin == "" {
+		// Default to external origin
+		newRepoReq.Origin = pointy.Pointer(config.OriginExternal)
+	}
+
 	ApiFieldsToModel(newRepoReq, &newRepoConfig, &newRepo)
 
-	cleanedUrl := models.CleanupURL(newRepo.URL)
-	if err := r.db.WithContext(ctx).Where("url = ?", cleanedUrl).FirstOrCreate(&newRepo).Error; err != nil {
-		return api.RepositoryResponse{}, DBErrorToApi(err)
+	if newRepo.URL == "" {
+		if err := r.db.WithContext(ctx).Create(&newRepo).Error; err != nil {
+			return api.RepositoryResponse{}, DBErrorToApi(err)
+		}
+	} else if newRepo.URL != "" {
+		cleanedUrl := models.CleanupURL(newRepo.URL)
+		// Repo configs with the same URL share a repository object
+		if err := r.db.WithContext(ctx).Where("url = ?", cleanedUrl).FirstOrCreate(&newRepo).Error; err != nil {
+			return api.RepositoryResponse{}, DBErrorToApi(err)
+		}
```

Comment:
If a repository exists (not a repository config) and I create an upload repo with the URL of the existing repository, I get no error, because it fetches the existing repository (which would have a non-upload origin). If the repository does not exist, I get a 400 error because "an upload repo cannot contain a URL". I think I'd expect to get the same 400 in the first scenario.

Reply:
updated and add a test for create and update for this situation